### ISSUE TITLE
Add Docker image, Docker Compose example cluster, and README docs - [merged]

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1852]
_Merges add-docker -> master_

Closes #822 

~Blocked by https://earth.bsc.es/gitlab/es/autosubmit/-/issues/1106~

**Comment by kinow** [comment]
added 1 commit

<ul><li>dc794a51 - Our containers need --init for zombie jobs</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/364/diffs?diff_id=15940&start_sha=b4713d8a66db73b53b628a72c22daa83c9f20da8)

**Comment by kinow** [comment]
marked this merge request as **ready**

**Comment by kinow** [comment]
requested review from @mcastril

**Comment by kinow** [comment]
@mcastril we have a container! :tada: 

Tested and it can run a dummy experiment. There is also a recipe for a cluster with 2 worker nodes, but I didn't have time to create an example experiment.

We can discuss over the week, but I think this should be enough for what we need for DTO.

Cheers

**Comment by kinow** [comment]
In GitLab by @GinkaVanThielen on Aug 17, 2023, 11:49

mentioned in issue digital-twins/edito/project_management#18

**Comment by kinow** [comment]
requested review from @manuel-g-castro and removed review request for @mcastril

**Comment by kinow** [comment]
Can you review it, please, @manuel-g-castro ?

**Comment by kinow** [comment]
My env @manuel-g-castro :

```bash
kinow@ranma:~/Development/python/workspace$ uname -a
Linux ranma 5.15.0-78-generic #85-Ubuntu SMP Fri Jul 7 15:25:09 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
kinow@ranma:~/Development/python/workspace$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.3 LTS"
kinow@ranma:~/Development/python/workspace$ docker --version
Docker version 24.0.5, build ced0996
kinow@ranma:~/Development/python/workspace$ docker compose version
Docker Compose version v2.20.2
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 18, 2023, 10:25

Sorry Bruno, I hadn't realized that I was the original reviewer.

With the porting of the simulations to MN4 and the team absences, I am lagging in some notifications.

I'll try to provide some feedback too.

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 18, 2023, 10:46

Commented on [dockerfiles/Dockerfile line 171](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..dc794a513272ae5dec5a5562c21ecbfba6eb5857#diff-b165692dd91c825b3d80951bf15844e5R171)



I couldn't run commands directly from the arguments on the docker run. For example `docker run --rm -it kinow/autosubmit:latest autosubmit --version` was only restarting the ssh service. 

Upon checking the [docker reference](https://docs.docker.com/engine/reference/builder/#:~:text=/[Autosubmit 3.7.7](https://github.com/BSC-ES/autosubmit/milestone/20)CMD%20combinations%3A-,No%20ENTRYPOINT,-ENTRYPOINT%20exec_entry%20p1_entry), I think we need to make the entry point a command and not shell.

I created a shell script to be run as entrypoint to then be copied within the container.

```bash
#!/bin/bash

sudo service ssh restart

exec "$@"
```

```suggestion:-1+0
COPY entrypoint.sh /usr/local/bin/entrypoint.sh
ENTRYPOINT ["/usr/local/bin/./entrypoint.sh"]
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 18, 2023, 10:47

Commented on [dockerfiles/Dockerfile line 160](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..63615310cf7763157450852b1f924816bf896cac#diff-b165692dd91c825b3d80951bf15844e5R160)



My build failed here with:

```
Sending build context to Docker daemon  17.92kB
Error response from daemon: dockerfile parse error line 160: unknown instruction: APPLICATION/PDF=XPDF.DESKTOP
```

This is with Ubuntu 20.04 and Docker 20.10. I'll try later in my desktop where I have Ubuntu 22.04 and a more recent Docker version.

In any case, I guess the default will no work if `xpdf` is not installed, like in y case. And we have the xdg-open issue... anyway to adopt here the strategy of defining the DE variable?

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 18, 2023, 11:15

Commented on [dockerfiles/README.md line 104](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-906c9a466ab4cfc17a271318b1ea3d6eR104)



Here we need to tell autosubmit where he should look for the databases. To solve this I binded the `.autosubmitrc` file to the mamba user's home to the places where we want to store the dabase from whitin the container. We could also just create it during the build of the image.

```
[database]
path = /app/autosubmit/database
filename = autosubmit.db

[local]
path = /app/autosubmit

[globallogs]
path = /app/autosubmit/logs

[structures]
path = /app/autosubmit/metadata/structures

[historicdb]
path = /app/autosubmit/metadata/data

[historiclog]
path = /app/autosubmit/metadata/logs

[autosubmitapi]
url = http://192.168.11.91:8081 # Replace me?

```

Then, when creating the folders in the host, we need to make sure that **all users** have permission to write on it. Otherwise autosubmit will complain.

Also I binded everything that says [in the docs](https://autosubmit.readthedocs.io/en/latest/installation/index.html). Maybe we don't need to have all of that to make autosubmit work properly. So every run command is replaced by 

```bash
docker run --rm \               
  -v /tmp/autosubmit/database:/app/autosubmit/database \
  -v /tmp/autosubmit/logs:/app/autosubmit/logs \
  -v /tmp/autosubmit/metadata:/app/autosubmit/metadata \
  -v /home/mgimenez/Documents/local/dev/autosubmit/dockerfiles/autosubmitrc:/home/autosubmit/.autosubmitrc \
  kinow/autosubmit:latest \
  autosubmit command params
```

**Comment by kinow** [comment]
>This is with Ubuntu 20.04 and Docker 20.10. I'll try later in my desktop where I have Ubuntu 22.04 and a more recent Docker version.

Exactly. The heredoc in Dockerfile is from a recent version.

Your version is not too old. The last Docker 20.10 release is from April this year. But looks like they are moving quite fast.

>In any case, I guess the default will no work if `xpdf` is not installed, like in y case. And we have the xdg-open issue... anyway to adopt here the strategy of defining the DE variable?

Interestingly it worked for me even though I don't have xpdf. This X forward with Docker might work a little different than with SSH, I think. If we fix the syntax error we can confirm if that works for you without the need of the variable.

If you are not able to try with a newer version we can ask @manuel-g-castro to verify if the `xpdf` works for him, and whether his host computer has `xpdf` or not (:pray: )

Thanks @mcastril !

**Comment by kinow** [comment]
>to the mamba user's home

Ah, the image is configured to run with the `autosubmit` user instead of `mamba`.

See `ARG NEW_MAMBA_USER=autosubmit` on line 53. The `.autosubmitrc` should be already properly configured on that user's home.

Can you try with that user? Easier to test with `docker compose up` :grimacing:

**Comment by kinow** [comment]
Good catch, I didn't test that. Let me try your suggestion.

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 18, 2023, 14:02

Commented on [dockerfiles/README.md line 104](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-906c9a466ab4cfc17a271318b1ea3d6eR104)



Sorry I wasn't clear! It is the autosubmit user, but set as MAMBA_USER. I have tested it without changing it.

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 18, 2023, 14:58

Commented on [dockerfiles/Dockerfile line 160](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..63615310cf7763157450852b1f924816bf896cac#diff-b165692dd91c825b3d80951bf15844e5R160)



Now I could build the Dockerfile (using Docker v24.10). I think that you should add a note regarding the minimal Docker version in the README. The latest Ubuntu long-term release is 22.04, and it still has Docker 20.10 in its repos.  

I don't have xpdf in this computer either, I'll report if I see any issue about it.

**Comment by kinow** [comment]
changed this line in [version 3 of the diff](/gitlab/es/autosubmit/-/merge_requests/364/diffs?diff_id=16191&start_sha=dc794a513272ae5dec5a5562c21ecbfba6eb5857#1e818c5a1411960e8f98fa04996e27480f047a65_171_175)

**Comment by kinow** [comment]
added 2 commits

<ul><li>4090b5b2 - Add entrypoint (thanks @manuel-g-castro)</li><li>2867e185 - Add prerequisites to containers README (thanks @mcastril)</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/364/diffs?diff_id=16191&start_sha=dc794a513272ae5dec5a5562c21ecbfba6eb5857)

**Comment by kinow** [comment]
>I have tested it without changing it.

And it worked?

I just did

```bash
$ docker build   -t kinow/autosubmit:4.0.84-bullseye-slim   -t kinow/autosubmit:latest   .
$ docker run --rm -it kinow/autosubmit:latest /bin/bash
(base) autosubmit@213d8a16bfd6:/app/autosubmit$ cat ~/.autosubmitrc 
[database]
path = /app/autosubmit//database/
filename = autosubmit.db

[local]
path = /app/autosubmit//experiments/

[globallogs]
path = /app/autosubmit//experiments/logs

[structures]
path = /app/autosubmit//experiments/metadata/structures

[historicdb]
path = /app/autosubmit//experiments/metadata/data

[historiclog]
path = /app/autosubmit//experiments/metadata/logs

[autosubmitapi]
url = http://192.168.11.91:8081 # Replace me?

(base) autosubmit@213d8a16bfd6:/app/autosubmit$ 
```

>Here we need to tell autosubmit where he should look for the databases. To solve this I binded the `.autosubmitrc` file to the mamba user's home to the places where we want to store the dabase from whitin the container. We could also just create it during the build of the image.

I think there is no need to update the `.autosubmitrc` after starting the container, as the `/home/autosubmit/.autosubmitrc` file is created during the `docker build`?

**Comment by kinow** [comment]
>I think that you should add a note regarding the minimal Docker version in the README.

+1! Done, thanks!

>I don't have xpdf in this computer either, I'll report if I see any issue about it.

Thank you :bow:

**Comment by kinow** [comment]
Just added an entrypoint file and updated CMD to be an array too. Now passing nothing to the run/exec should result in `ENTRYPOINT + CMD`, or passing something will replace `CMD`. I used your `exec $@` approach too. Thanks!

**Comment by kinow** [comment]
```bash
kinow@ranma:~/Development/python/workspace/autosubmit/dockerfiles$ docker run --rm -it kinow/autosubmit:latest autosubmit --version
Restarting OpenBSD Secure Shell server: sshd.
4.0.84

```

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 18, 2023, 15:20

Commented on [dockerfiles/README.md line 104](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-906c9a466ab4cfc17a271318b1ea3d6eR104)



It worked, but now I am not sure what was the issue :thinking:

**Comment by kinow** [comment]
added 1 commit

<ul><li>56db9ec4 - Set the context to the dockerfiles directory</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/364/diffs?diff_id=16193&start_sha=2867e185de9faccf024eebd96e7c0caf39452ec0)

**Comment by kinow** [comment]
added 3 commits

<ul><li>8c8121eb - Add entrypoint (thanks @manuel-g-castro)</li><li>59a74c9a - Add prerequisites to containers README (thanks @mcastril)</li><li>a17f79c4 - Set the context to the dockerfiles directory</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/364/diffs?diff_id=16195&start_sha=56db9ec4db050c1ce682f584c65d5b6cc776fc3b)

**Comment by kinow** [comment]
Hmmm, are you on wayland or X server?

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 18, 2023, 15:51

Commented on [dockerfiles/Dockerfile line 160](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..63615310cf7763157450852b1f924816bf896cac#diff-b165692dd91c825b3d80951bf15844e5R160)



I am user docker compose. But in both cases it fails with to open x11.

Nop, it is not working even setting the display environment variable... Will see if it works with [the network flag](https://askubuntu.com/questions/1245518/how-to-install-xpdf-on-ubuntu-20-04) that the people from stackoverflow mention.

**Comment by kinow** [comment]
```bash
$ loginctl
SESSION  UID USER  SEAT  TTY
     c2 1000 kinow seat0 

1 sessions listed.
$ loginctl show-session c2 -p Type
Type=x11
```

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 18, 2023, 15:53

Commented on [dockerfiles/docker-compose.yml line 23](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..63615310cf7763157450852b1f924816bf896cac#diff-d8a76c9d329c131d166f2bb0a12745f0R23)



```
SESSION  UID USER     SEAT  TTY 
      2 1001 mgimenez seat0 tty2
     c1  125 gdm      seat0 tty1

2 sessions listed.
mgimenez@ bsces107930 ~ % loginctl show-session 2 -p Type           

Type=x11
```

**Comment by kinow** [comment]
Weird. Without this `RUN mkdir /tmp/.X11-unix/` building and running with the `-v` lines above does not work then?

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 18, 2023, 15:56

Commented on [dockerfiles/docker-compose.yml line 23](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..63615310cf7763157450852b1f924816bf896cac#diff-d8a76c9d329c131d166f2bb0a12745f0R23)



with the -v it worked when executing from the `docker run`. But on the compose it failed...

**Comment by kinow** [comment]
I tried to avoid that `--net host` (I think that's the flag you are testing?) as that's not recommended due to security issue. In our case that should be fine, but in Mercator where we are supposed to deploy this image it's better if we can avoid it, I think. But curious to know if that solves the issue.

**Comment by kinow** [comment]
Strange, because under the `autosubmit` version there should be entries under `environment` and `volumes` that should produce the same as in `docker run` :dizzy_face:

Maybe try a `docker compose down` or `docker compose rm`, and then `docker compose build`. (Or nuke everything or images and containers created before testing again...)

**Comment by kinow** [comment]
@mcastril I updated the `Dockerfile` with your suggestion and tested, and it appears to work fine. But I started searching about this one

>It's important to protect the variable `$\@`

And I couldn't find an explanation for it. Could you explain this one, please? :slight_smile: I enjoy when I can learn something in code reviews like this!!! Thanks

**Comment by kinow** [comment]
Calling it a day, will look at the comments and try to update the branch based on any feedback that I haven't replied. Thanks **a lot** @manuel-g-castro and @mcastril ! I am sure MOi/DTO and other future users will appreciate the thorough review & testing we are doing on this new docker files :bow:  Have a great weekend! :wave: :walking:

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 18, 2023, 16:03

Commented on [dockerfiles/README.md line 69](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-906c9a466ab4cfc17a271318b1ea3d6eR69)



I don't have so much experience as you with containers @kinow 

I am curious about why we cannot use `RUN --mount` in these situations

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 18, 2023, 16:04

Commented on [dockerfiles/Dockerfile line 167](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..757e2abaa472c7dc9e6fc18e2d978449892a4f20#diff-b165692dd91c825b3d80951bf15844e5R167)



```suggestion:-0+0
COPY entrypoint.sh /usr/local/bin/entrypoint.sh
COPY <<EOF /usr/local/bin/entrypoint.sh
#!/bin/bash
sudo service ssh restart
exec "$\@"
EOF

```

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 18, 2023, 16:11

Commented on [dockerfiles/README.md line 104](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-906c9a466ab4cfc17a271318b1ea3d6eR104)



I confirm the steps also worked for me:

```
(base) autosubmit@49ba651db5ff:/app/autosubmit$ cat ~/.autosubmitrc 
[database]
path = /app/autosubmit//database/
filename = autosubmit.db

[local]
path = /app/autosubmit//experiments/

[globallogs]
path = /app/autosubmit//experiments/logs

[structures]
path = /app/autosubmit//experiments/metadata/structures

[historicdb]
path = /app/autosubmit//experiments/metadata/data

[historiclog]
path = /app/autosubmit//experiments/metadata/logs

[autosubmitapi]
url = http://192.168.11.91:8081 # Replace me?
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 18, 2023, 16:18

Commented on [dockerfiles/README.md line 118](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-906c9a466ab4cfc17a271318b1ea3d6eR118)



I think this is assuming that the command is run from $HOME isn't it ?

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 18, 2023, 16:27

Commented on [dockerfiles/Dockerfile line 160](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..63615310cf7763157450852b1f924816bf896cac#diff-b165692dd91c825b3d80951bf15844e5R160)



Oh, once I ran `create` I saw a nice image popping up, and I don't have `xpdf` installed :smile_cat: 

```
Saving unified data..
Saving experiment data into /app/autosubmit/experiments/a000/conf/metadata


Plotting the jobs list...
Plotting...
Plot created at /app/autosubmit//experiments/a000/plot/a000_20230818_1426.pdf
Writing status txt...
Status txt created at <_io.TextIOWrapper name='/app/autosubmit//experiments/a000/status/a000_20230818_1426.txt' mode='w+' encoding='UTF-8'>
Warning: Cannot convert string "-*-helvetica-medium-r-normal--12-*-*-*-*-*-iso8859-1" to type FontStruct
Warning: Cannot convert string "-*-courier-medium-r-normal--12-*-*-*-*-*-iso8859-1" to type FontStruct
Warning: Cannot convert string "-*-times-bold-i-normal--20-*-*-*-*-*-iso8859-1" to type FontStruct
Warning: Cannot convert string "-*-times-medium-r-normal--16-*-*-*-*-*-iso8859-1" to type FontStruct
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 18, 2023, 16:32

Commented on [dockerfiles/README.md line 129](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..760d1cbbd27714723e8d157596efe6ce29292e8e#diff-906c9a466ab4cfc17a271318b1ea3d6eR129)



That's what we already do in the previous sections (Use volumes). We can change the order of the sections, or make a reference.

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 18, 2023, 16:34

Commented on [dockerfiles/README.md line 145](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..760d1cbbd27714723e8d157596efe6ce29292e8e#diff-906c9a466ab4cfc17a271318b1ea3d6eR145)



The `keygen.sh` tool gives options (ecdsa, rsa, RSA bits...)

Maybe say that we'll use `rsa`  ?

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 18, 2023, 16:51

Commented on [dockerfiles/README.md line 186](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..760d1cbbd27714723e8d157596efe6ce29292e8e#diff-906c9a466ab4cfc17a271318b1ea3d6eR186)



I couldn't make it work.

I have:

```
NAME                           IMAGE                                       COMMAND                  SERVICE             CREATED             STATUS              PORTS
autosubmit                     dockerfiles-autosubmit                      "sudo /usr/sbin/sshd…"   autosubmit          9 minutes ago       Up 7 minutes        0.0.0.0:32771->22/tcp, :::32771->22/tcp
dockerfiles-computing_node-1   lscr.io/linuxserver/openssh-server:latest   "/init"                  computing_node      9 minutes ago       Up 7 minutes        2222/tcp, 0.0.0.0:32770->22/tcp, :::32770->22/tcp
```

```
~/tmp/autosubmit/dockerfiles$ sudo docker exec -ti dockerfiles-autosubmit-1 /bin/bash
Error response from daemon: No such container: dockerfiles-autosubmit-1
```

```
~/tmp/autosubmit/dockerfiles$ sudo docker exec -ti dockerfiles-autosubmit /bin/bash
Error response from daemon: No such container: dockerfiles-autosubmit
```

```
~/tmp/autosubmit/dockerfiles$ sudo docker exec -ti -d dockerfiles-autosubmit /bin/bash
Error response from daemon: No such container: dockerfiles-autosubmit
```

```
~/tmp/autosubmit/dockerfiles$ sudo docker exec -ti -d autosubmit /bin/bash
```

The last one does not complain, but does nothing

```
~/tmp/autosubmit/dockerfiles$ sudo docker exec -ti -d autosubmit autosubmit -v
```

Same result

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 18, 2023, 16:54

Commented on [dockerfiles/docker-compose.yml line 14](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-d8a76c9d329c131d166f2bb0a12745f0R14)



I changed them to `/tmp/autosubmit/database` and `/tmp/autosubmit/experiments` as they were the ones that we were using above

**Comment by kinow** [comment]
The folders under `/tmp/autosubmit` were used to run the container with a [bind mount](https://docs.docker.com/storage/bind-mounts/) (basically a directory in your computer being bound to a directory inside the container). I used `/tmp` to avoid users testing the containers to have to remember to remove directories used in the containers.

The Docker Compose uses [named volumes](https://docs.docker.com/storage/volumes/). I added this example to have a way to test everything with just `docker compose up`, ready to be used by anyone (not only development).

When you run Docker Compose the first time it will start the cluster and create the named volumes. Then the next times you start the cluster again, the volumes will be re-used. They are available normally under `/var/lib/docker/volumes` and are managed by Docker.

I thought about having identical setup for Docker and for Docker Compose, but to have named volumes it is a bit annoying that before you `docker run` you have to remember to `docker volume create`.

Do you think we need to update our docs, or modify the examples or the Dockerfile here?

**Comment by kinow** [comment]
Ah, that might be because I updated the context in the `docker-compose.yml`, sorry. I tested everything, and wrote a `test.sh` script that validates the commands used in the dockerfiles/README.md file.

We can include any commands you or @manuel-g-castro  find useful there. This way I will remember what to test if the images are modified (for example, I had forgotten to test the `docker run ... autosubmit version`, and Manuel found it while reviewing this MR.)

Also added a note that those commands can be turned into a proper test later, when our CICD workers have support to Docker.

**Comment by kinow** [comment]
I also added more notes in the README.md, including about the container names generated by Docker Compose.

I think depending on how many clusters you start, Docker Compose may add/increase the suffix, creating `autosubmit-1`, `autosubmit-2`, or `dockerfiles-autosubmit1`, etc. (by default, it's just `autosubmit` for the AS container).

For the computing nodes, they will always be something like `dockerfiles-computing_node-N`, where `N` will go from `1` to as many nodes as you create (if you have already created `2`, then it starts at `3`, and so it goes).

**Comment by kinow** [comment]
Added a note about it :thumbsup:

**Comment by kinow** [comment]
Done! Also updated the paragraph above to end with "... all your jobs and logs will be
deleted if you have not used volumes.". Thank you!

**Comment by kinow** [comment]
Ah, I wrote it after creating the SSH keys with the keygen command. Good catch. I've moved the part about the keygen from Docker Compose to the Prerequisites.

I left a note in the Docker Compose section to refer to that, and added another one above this command.

**Comment by kinow** [comment]
I think I had pushed it almost at the same time you suggested this change (I think I read another comment you left).

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/c496f637b5a2f794a0cf46c597f9974b/image.png)

**Comment by kinow** [comment]
>I don't have so much experience as you with containers @kinow

I am sure you'll learn anything I know about containers that you are not already aware in a very short time, and probably teach me something new very soon :sweat_smile: The Docker architecture & principles involve a lot of Linux, shell, and environment isolation & portability (which you will probably find a lot of similarities with things from your experience with HPC).

>I am curious about why we cannot use `RUN --mount` in these situations

You mean inside the `Dockerfile`? If so, the `RUN` is used there to bind volumes during the build stage only (although I think I never used that).

If you mean the `docker run --mount`, the `-v` is equivalent to`--mount type=bind,source=...,destination=...`. Just simpler. I think I used `--mount` a couple of times so far (maybe for readonly?). But otherwise I always use `-v`.

Depending on the platform your are deploying your containers, you may never use these settings. AWS, Kubernetes, Google Cloud, have their own solution to orchestrate containers and manage volumes, and normally use Web UI or some YAML or equivalent to configure it (`docker compose` is the simple/closest example, although very much like `docker`).

**Comment by kinow** [comment]
I've updated the README and added a `test.sh` file @manuel-g-castro . Hopefully that will help during the review, and also to help me to avoid accidentally breaking it in future commits :grimacing:

**Comment by kinow** [comment]
Great news! And thinking now with a fresh mind, I think I understand why this works.

With our normal `autosubmit create` in our VM's, we have the X client communicating with the X server via SSH -- via networking, with a x server proxy listener controlling the application that must exist on both ends.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/af1fd2109e9d6b20c41b13e744da3604/image.png)

But with `autosubmit create` inside the container, the container is running an X client that has access to the X11 files (via bound volumes), and acts as if using a local display -- no networking, similar to how it's done on local computers.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/a9981632585c29ed63b08dca59682aef/image.png)

So with Docker, `xpdf` starts and communicates via X11 using the unix socket bound via volumes, and tells the X server to display the xpdf application. Since the unix socket is bound inside the container, linux & namespaces are not preventing the communication.

And thinking better now, that's probably not very safe.

I found this utility, [x11docker](https://github.com/mviereck/x11docker), that offers a secure way to run GUI applications in Docker. Their first point about security:

>Runs a second X server to avoid X security leaks.
>
>>    This in opposite to widespread solutions that share host X socket of display :0, thus breaking container isolation, allowing keylogging and remote host control. (However, x11docker provides this with fallback option --hostdisplay).

@mcastril , I think we should update the Docker/Compose/readme files to leave commented the steps to have the `xpdf` and `autosubmit create` working, but leave the default containers X-less. To avoid any security issues in our infra, or in EDITO. WDYT?

Images from: https://goteleport.com/blog/x11-forwarding/

**Comment by kinow** [comment]
changed this line in [version 7 of the diff](/gitlab/es/autosubmit/-/merge_requests/364/diffs?diff_id=16209&start_sha=760d1cbbd27714723e8d157596efe6ce29292e8e#b23687ffaf279d4d1dd18aa9bb9334d3da0b6851_129_145)

**Comment by kinow** [comment]
changed this line in [version 7 of the diff](/gitlab/es/autosubmit/-/merge_requests/364/diffs?diff_id=16209&start_sha=760d1cbbd27714723e8d157596efe6ce29292e8e#b23687ffaf279d4d1dd18aa9bb9334d3da0b6851_145_149)

**Comment by kinow** [comment]
changed this line in [version 7 of the diff](/gitlab/es/autosubmit/-/merge_requests/364/diffs?diff_id=16209&start_sha=760d1cbbd27714723e8d157596efe6ce29292e8e#b23687ffaf279d4d1dd18aa9bb9334d3da0b6851_186_197)

**Comment by kinow** [comment]
added 1 commit

<ul><li>7c9513be - Add Docker image, Docker Compose example cluster, and README docs</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/364/diffs?diff_id=16209&start_sha=760d1cbbd27714723e8d157596efe6ce29292e8e)

**Comment by kinow** [comment]
added 111 commits

<ul><li>7c9513be...f5677a50 - 110 commits from branch <code>master</code></li><li>d5b20099 - Add Docker image, Docker Compose example cluster, and README docs</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/364/diffs?diff_id=16210&start_sha=7c9513be321471b325a90a963ee0f046278ce2ed)

**Comment by kinow** [comment]
Rebased and squashed.

**Comment by kinow** [comment]
mentioned in issue digital-twins/edito/project_management#17

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 10:16

Commented on [dockerfiles/Dockerfile line 171](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..dc794a513272ae5dec5a5562c21ecbfba6eb5857#diff-b165692dd91c825b3d80951bf15844e5R171)



This is usually done to avoid expanding the variable when the HEREDOC is generated so that the placeholder is preserved intact.

I was not sure it was needed in this case (I didn't know if the Docker runtime would expand Bash variables), but the recipe did not work until I added the slash. However, now I note that the slash is not well positioned (it should be placed before the dollar sign). 

I think that we should test it again without the slash, maybe it was something else

**Comment by kinow** [comment]
>This is usually done to avoid expanding the variable when the HEREDOC is generated so that the placeholder is preserved intact.

Ah! That makes sense! Did some local tests and you are correct in both escaping it, and that the `\` must come before the `$`.

I just updated it and re-run `test.sh` (even though it worked yesterday with the previous example too, strange...)

Thanks!

**Comment by kinow** [comment]
added 1 commit

<ul><li>63615310 - Fix variable escape</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/364/diffs?diff_id=16212&start_sha=d5b20099298c8a9f537d842643451219a2049a1c)

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 21, 2023, 10:52

Commented on [dockerfiles/README.md line 104](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-906c9a466ab4cfc17a271318b1ea3d6eR104)



Just a clarification. The issue I was having was due to permissions on the host's files. I solved this by adding a `chmod a+rw -R /tmp/autosubmit/` to the setup file that Bruno provided to us. Maybe giving all of those permissions is not necessary, but it works.

I needed the weekend to figure this out.

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 21, 2023, 11:06

Commented on [dockerfiles/docker-compose.yml line 23](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..63615310cf7763157450852b1f924816bf896cac#diff-d8a76c9d329c131d166f2bb0a12745f0R23)



Okay! The script works! Although there is a small little nuisance. On the `setup()` function, if you have files under the `/tmp/autosubmit/` created by a previous run of the container, it will fail to remove them, since the were created by docker. You need privileged permissions to remove it.

**Comment by kinow** [comment]
>Okay! The script works!

:tada: 

>Although there is a small little nuisance. On the `setup()` function, if you have files under the `/tmp/autosubmit/` created by a previous run of the container, it will fail to remove them, since the were created by docker. You need privileged permissions to remove it.

I wonder if we have something different, our Docker versions, or Ubuntu/Linux flavor, some file system setting... because it works for me™ no matter how many times I run it.

```bash
# start the script
kinow@ranma:/tmp/autosubmit$ ls -lah
total 1.9M
drwxrwxr-x  4 kinow kinow 4.0K Aug 21 11:22 .
drwxrwxrwt 25 root  root  1.9M Aug 21 11:22 ..
drwxrwxr-x  2 kinow kinow 4.0K Aug 21 11:22 database
drwxrwxr-x  2 kinow kinow 4.0K Aug 21 11:22 experiments
# once it succeeds
kinow@ranma:/tmp$ ls /tmp/autosubmit
ls: cannot access '/tmp/autosubmit': No such file or directory
```

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 21, 2023, 11:53

Commented on [dockerfiles/docker-compose.yml line 23](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..63615310cf7763157450852b1f924816bf896cac#diff-d8a76c9d329c131d166f2bb0a12745f0R23)



Maybe the files that are generated do not have the correct permissions to be removed...

```
% ./test.sh 
+ trap handle_errors ERR
+ setup
+ rm -rf /tmp/autosubmit/
rm: cannot remove '/tmp/autosubmit/experiments/logs/20230821_084938_delete_a000_err.log': Permission denied
rm: cannot remove '/tmp/autosubmit/experiments/logs/20230821_084935_describe_*_err.log': Permission denied
rm: cannot remove '/tmp/autosubmit/experiments/logs/20230821_084933_expid.log': Permission denied
rm: cannot remove '/tmp/autosubmit/experiments/logs/20230821_084938_delete_a000.log': Permission denied
rm: cannot remove '/tmp/autosubmit/experiments/logs/20230821_084935_describe_*.log': Permission denied
rm: cannot remove '/tmp/autosubmit/experiments/logs/20230821_084933_expid_err.log': Permission denied
```

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 21, 2023, 12:15

Commented on [dockerfiles/docker-compose.yml line 23](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..63615310cf7763157450852b1f924816bf896cac#diff-d8a76c9d329c131d166f2bb0a12745f0R23)



:tada: I managed to open the `xpdf` GUI from the container! :tada: 

But maybe not in a pretty/secure way. I has to run on the host `xhost local:root` prior to run the docker image.

By any chance have you done that, @kinow? Or maybe it is on or `.bashrc` (which I wouldn't know why)?

**Comment by kinow** [comment]
> I managed to open the `xpdf` GUI from the container!

:tada: 

>But maybe not in a pretty/secure way. I has to run on the host `xhost local:root` prior to run the docker image.
>
>By any chance have you done that, @kinow? Or maybe it is on or `.bashrc` (which I wouldn't know why)?

It might be due to how your X server is configured, or maybe the docker versions or some other environment setting.

But see the thread above this one on GitLab, where I mentioned to @mcastril that I think it might be better (and more secure) to disable the X redirect & remove xpdf, leaving the image completely headless.

Still good to know that it works - even though not very safe - in case we ever need to run some GUI application in a container :slight_smile: 

Thanks @manuel-g-castro !

**Comment by kinow** [comment]
@manuel-g-castro :point_up:

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 21, 2023, 12:54

Commented on [dockerfiles/docker-compose.yml line 23](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..63615310cf7763157450852b1f924816bf896cac#diff-d8a76c9d329c131d166f2bb0a12745f0R23)



Then, I guess we can resolve the thread.

Well, I saw that there is the option of creating a vnc server or forwarding X via ssh.

**Comment by kinow** [comment]
>Then, I guess we can resolve the thread.

:thumbsup: 

>Well, I saw that there is the option of creating a vnc server or forwarding X via ssh.

Hmm, that does sound like a safer and more portable option. If that works with no security issues, we could enable that for the Edito DTO workshop (which will use these Docker images).

Although they could also decide to use a less safe approach there... probably simpler to move it to a follow-up issue/discussion, I guess.

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 21, 2023, 12:56

Commented on [dockerfiles/README.md line 206](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-906c9a466ab4cfc17a271318b1ea3d6eR206)



If we have configured via docker compose isn't this port 22?

```suggestion:-0+0
$ ssh -i id_rsa -p 22 autosubmit@localhost
```

**Comment by kinow** [comment]
Not really, I believe.

In `docoker-compose.yml` we only declare that it uses the port `22`.

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/aeffa6bb6440a2bf3d0e0d18e11ec1f2/image.png)

Since we don't provide a host port, Docker Compose assigned a randomly generated port. That's why users must list or inspect containers to get the port number and use it here (the text above should be addressing that).

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 21, 2023, 13:05

Commented on [dockerfiles/README.md line 206](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-906c9a466ab4cfc17a271318b1ea3d6eR206)



Well, once that is addressed, I'll approve the merge.

**Comment by kinow** [comment]
>Well, once that is addressed, I'll approve the merge.

You mean the X server & xpdf?

Here we cannot really bind a port as if you start two clusters the second will fail. But a workflow running inside the cluster shouldn't have a problem connecting to `dockerfiles-compute_node-N`, port `2222` (i.e. this setup is made to have an internal/cluster network, shared by the nodes -- for host user, they can connect via the random port or `docker exec`...)

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 13:15

Commented on [dockerfiles/Dockerfile line 160](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..63615310cf7763157450852b1f924816bf896cac#diff-b165692dd91c825b3d80951bf15844e5R160)



So the container is using the `xpdf` package installed inside the container and directly using the local screen. 

I see this can be seen as potentially unsafe but I am not sure that I understand all the implications.

We are also in the context that we run the container locally, but how would it behave when you mix both scenarios?: running the container remotely and forwarding X11

**Comment by kinow** [comment]
>We are also in the context that we run the container locally, but how would it behave when you mix both scenarios?: running the container remotely and forwarding X11

Good point! Hadn't thought about that! It will probably fail... I will remove xpdf and the X11 forwarding from this merge request.

Later we can review if we will use a similar approach, or the vnc server @manuel-g-castro suggested below.

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 13:21

Commented on [dockerfiles/README.md line 186](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..760d1cbbd27714723e8d157596efe6ce29292e8e#diff-906c9a466ab4cfc17a271318b1ea3d6eR186)



Thanks Bruno, I'll retake the tests after the eFlows meeting.

**Comment by kinow** [comment]
Sounds good! I'm removing the xpdf and X stuff right now. A new commit should arrive in some minutes :timer:

**Comment by kinow** [comment]
changed this line in [version 10 of the diff](/gitlab/es/autosubmit/-/merge_requests/364/diffs?diff_id=16219&start_sha=63615310cf7763157450852b1f924816bf896cac#1e818c5a1411960e8f98fa04996e27480f047a65_160_157)

**Comment by kinow** [comment]
changed this line in [version 10 of the diff](/gitlab/es/autosubmit/-/merge_requests/364/diffs?diff_id=16219&start_sha=63615310cf7763157450852b1f924816bf896cac#74fc6feb8c0c353caa01966b68556a32ced8507a_23_20)

**Comment by kinow** [comment]
added 1 commit

<ul><li>eed1ba92 - Add Docker image, Docker Compose example cluster, and README docs</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/364/diffs?diff_id=16219&start_sha=63615310cf7763157450852b1f924816bf896cac)

**Comment by kinow** [comment]
Done. Reviewed the files in this MR, replacing existing PDF/X code/docs by comments. Tested with `test.sh`, squashed commits.

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 21, 2023, 14:00

Commented on [dockerfiles/README.md line 206](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-906c9a466ab4cfc17a271318b1ea3d6eR206)



I mean just correcting the docs to inform that you have to pass the port as specified by docker.

**Comment by kinow** [comment]
There is some text already, do you think we need to improve that part?

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/cb68b553e5dd07b6c441413adc1e687d/image.png)

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 21, 2023, 14:05

Commented on [dockerfiles/README.md line 206](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-906c9a466ab4cfc17a271318b1ea3d6eR206)



No. It is fine like this.

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 15:35

Commented on [dockerfiles/README.md line 69](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-906c9a466ab4cfc17a271318b1ea3d6eR69)



You are right Bruno, we cannot use RUN for this !

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 15:51

Commented on [dockerfiles/Dockerfile line 171](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..dc794a513272ae5dec5a5562c21ecbfba6eb5857#diff-b165692dd91c825b3d80951bf15844e5R171)



I verified that the escape is necessary.

It should be `\$@` but, in this case, `$\@` is nothing that Bash can interpret either, and it worked the same...

```
mcastril@luxor:~/tmp/autosubmit/dockerfiles$ hello() { echo $@; }; hello "hola" "adios" 
hola adios
mcastril@luxor:~/tmp/autosubmit/dockerfiles$ hello() { echo \$@; }; hello "hola" "adios" 
$@
mcastril@luxor:~/tmp/autosubmit/dockerfiles$ hello() { echo $\@; }; hello "hola" "adios" 
$@
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 15:55

Commented on [dockerfiles/README.md line 186](https://github.com/BSC-ES/autosubmit/compare/09c3361d4a81d3972e1b5d1ee7a6029b9559ce22..760d1cbbd27714723e8d157596efe6ce29292e8e#diff-906c9a466ab4cfc17a271318b1ea3d6eR186)



I didn't update yet, but after changing the compose image:

`dockerfile: Dockerfile`

It worked for me.

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 16:05

Commented on [dockerfiles/docker-compose.yml line 14](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-d8a76c9d329c131d166f2bb0a12745f0R14)



Thanks, Bruno, for the enlightening explanation. I checked that, indeed, the volumes are created in `/var/lib/docker/volumes`

I would only suggest changing the comment:

`# Or map it to local folders.`

Made me think that these were pre-existing local folders 

Maybe changing it to

`# Using named volumes managed by Docker`

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 16:10

Commented on [dockerfiles/README.md line 206](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-906c9a466ab4cfc17a271318b1ea3d6eR206)



So, in this example, would it be `32783` ?

`autosubmit                     dockerfiles-autosubmit                      "sudo /usr/sbin/sshd…"   autosubmit          33 minutes ago      Up 32 minutes       0.0.0.0:32783->22/tcp, :::32783->22/tcp`

**Comment by kinow** [comment]
I think so, yes.

**Comment by kinow** [comment]
added 1 commit

<ul><li>147340c0 - Add Docker image, Docker Compose example cluster, and README docs</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/364/diffs?diff_id=16227&start_sha=eed1ba92b0e91a09eefba41f8485f78fe35de1d7)

**Comment by kinow** [comment]
Sounds good to me, thanks for the suggestion. Done!

**Comment by kinow** [comment]
xpdf and X removed. Resolving thread.

**Comment by kinow** [comment]
Updated comment as suggested by @mcastril , squashed commits, and marked as resolved some threads to make reviewing easier. Feel free to re-open any thread if not solved yet, @mcastril . Thanks!

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 16:22

Commented on [dockerfiles/README.md line 206](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-906c9a466ab4cfc17a271318b1ea3d6eR206)



I get `Failed publickey for autosubmit from 172.18.0.1 port 43410` on trying to connect

I thought that it was due to the public key was missing in `authorized_keys` but it didn't work either. I'm gonna see if I skipped any step.

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 16:31

Commented on [dockerfiles/README.md line 206](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..147340c063e64a6e0f736d88ae5005172e7d3b48#diff-906c9a466ab4cfc17a271318b1ea3d6eR206)



It works for me now, so surely I had messed it up

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 16:33

resolved all threads

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 16:34

Commented on [dockerfiles/docker-compose.yml line 23](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..63615310cf7763157450852b1f924816bf896cac#diff-d8a76c9d329c131d166f2bb0a12745f0R23)



Last test. Gonna test the updated script too.

**Comment by kinow** [comment]
Thanks for testing it Miguel!!

:drum:

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 16:50

Commented on [dockerfiles/docker-compose.yml line 23](https://github.com/BSC-ES/autosubmit/compare/f5677a50ffff8489b920f653ee68e4b9c1abaf4c..63615310cf7763157450852b1f924816bf896cac#diff-d8a76c9d329c131d166f2bb0a12745f0R23)



The `test.sh` script worked like a charm! Thank you so much Bruno!

Something we could do in the future is maybe to create an Autosubmit platform file for the docker cluster and try running a dummy workflow there.

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 16:50

resolved all threads

**Comment by kinow** [comment]
>The `test.sh` script worked like a charm! Thank you so much Bruno!

Great!

>Something we could do in the future is maybe to create an Autosubmit platform file for the docker cluster and try running a dummy workflow there.

Agreed! I tested a dummy workflow with `local` platform, but didn't try using multiple hosts with the cluster (although that should work, in theory...)

**Comment by kinow** [comment]
resolved all threads

**Comment by kinow** [comment]
Thanks for resolving the threads @mcastril ! Is there anything pending? I am just waiting for you to hit the approve button before merging :slight_smile:

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 17:29

approved this merge request

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 21, 2023, 17:30

Approved!

**Comment by kinow** [comment]
Hooray! Merging, thanks @mcastril and @manuel-g-castro !

**Comment by kinow** [comment]
resolved all threads

**Comment by kinow** [comment]
mentioned in commit 55d31b25a0a4606b9188a19c541b78250752f5ef

**Comment by kinow** [comment]
In GitLab by @mcastril on Aug 22, 2023, 17:19

Thanks for the good work @kinow !

Can you prepare an e-mail for MOI to share the container with them?

**Comment by kinow** [comment]
Thanks Miguel! And sure thing! Will prepare the email tomorrow :thumbsup:

**Comment by kinow** [comment]
resolved all threads
